### PR TITLE
[WIP] token selector: streamline filter interfaces #186

### PR DESCRIPTION
**Pull request by adecaro**
Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: token/selector.go [hunk 1]
```diff
@@ -25,21 +26,31 @@ var (
...
+	// ID returns the ID of the identity the selection should select for.
 	ID() string
+}
+
+// SelectorFilterByUnspentToken is a SelectorFilter that selects tokens that are not yet spent.
+type SelectorFilterByUnspentToken interface {
+	SelectorFilter
 	// ContainsToken returns true if the passed token is recognized, false otherwise.
 	ContainsToken(token *token2.UnspentToken) bool
 }
```
Comment (mbrandenburger) [inline token/selector.go:44]: Mhhh. The filter used here is only there to dispatch the select call to either to selectByOwnerID or selectByFilter.

My understanding of a filter is something that once applies on a data structure and there is a predicate to be true in order to get the item or not. this is not the case with the selectorFilter. Also, the filter may have more predicates. For instance, tokenType could be one. We maybe wanna add more filters, like minTokenSize, maxTokenSize, oldestToken, …. In order to model is nicely, we could throw away the filter interface it is right now and make it just a function, or a list of functions.  WDYT?

**Review by mbrandenburger — APPROVED** [review]
This looks already better than before. See my comment on the filter, but let's address this later.

File: token/selector.go (hunk 1 above)
Comment (mbrandenburger) [reply token/selector.go:44]: @adecaro is the ID of the identity always a string? In some code I saw that this requires to do something some casting of the actual identity using the "raw material" `string(Owner.Raw)`.

File: token/services/selector/selector.go [hunk 2]
```diff
@@ -7,11 +7,11 @@ SPDX-License-Identifier: Apache-2.0
 package selector
 
 import (
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
 	"strconv"
 	"time"
 
 	"github.com/hashicorp/go-uuid"
-	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
 	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/services/selector/selector.go:15]: if I run goimports, it gets worse 😆

**Comment by adecaro** [comment]
I will withdraw this waiting for a deeper discussion on the selector API :)

**Pull request closed without merging**